HotSpot/View Controllers/BookingViewController.m

```diff
@@ -83,4 +100,196 @@ - (IBAction)confirmClicked:(id)sender {
+    }
+    return cell;
+}
+
+- (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section { 
+    return 24 * 4;
+}
+
+- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
+    if (self.timeSlots[indexPath.item].available) {
+        TimeCell *cell = [collectionView cellForItemAtIndexPath:indexPath];
+        NSDate *date = self.timeSlots[indexPath.item].date;
+        if ([startIndexPath isEqual:indexPath]) {
+            // repick start time
+            [self reset];
+        }
+        else if (pickingStartTime) {
+            startIndexPath = indexPath;
+            startTime = date;
+            pickingStartTime = NO;
+            self.timeSlots[indexPath.item].chosen = YES;
+            cell.backgroundColor = [UIColor colorWithRed:0 green:.4 blue:1.0 alpha:1.0];
+            [self.collectionView reloadData];
+        }
+        else {
+            for (int i = startIndexPath.item; i < indexPath.item; i++) {
+                if ( self.timeSlots[i].available == NO) {
+                    return;
+                }
+            }
+            if (startIndexPath.item > indexPath.item) {
+                // start date cannot be later than end date
+                return;
+            }
+            endTime = date;
+            for (int i = startIndexPath.item; i < self.timeSlots.count; i++) {
+                if (i<=indexPath.item) {
+                    self.timeSlots[i].chosen = YES;
+                }
+                else {
+                    self.timeSlots[i].chosen = NO;
+                }
+                [self.collectionView reloadData];
+            }
+        }
+    }
+    
+}
+
+- (void)updateCells {
+    PFRelation *relation = [self.listing relationForKey:@"unavailable"];
+    PFQuery *query = relation.query;
+    [query orderByDescending:@"repeatsWeekly"];
+    
+    NSDate *date = self.datePicker.date;
+    NSInteger secondsPerDay = 60 * 60 * 24;
+    NSInteger timeSinceBeginningOfDay = (int)[date timeIntervalSinceReferenceDate] % secondsPerDay; // TODO: consider time zones
+    NSDate *beginningOfDay = [date dateByAddingTimeInterval: -timeSinceBeginningOfDay];
+    NSDate *endOfDay = [date dateByAddingTimeInterval: secondsPerDay];
+    
+    [query whereKey:@"endTime" greaterThan:beginningOfDay];
+    [query whereKey:@"startTime" lessThan:endOfDay];
+    
+    self.timeSlots = [NSMutableArray new];
+    for (NSInteger i = 0; i < 24 * 4; i ++) {
+        TimeSlot *newTimeSlot = [TimeSlot new];
+        [newTimeSlot setTime:i
+                    withDate:beginningOfDay];
+        [self.timeSlots addObject:newTimeSlot];
+    }
+    
+    [query findObjectsInBackgroundWithBlock:^(NSArray<TimeInterval *> *timeIntervals, NSError * _Nullable error) {
+        if (timeIntervals) {
+            for (TimeInterval *timeInterval in timeIntervals) {
```

Comment:
the logic here is complex, and should have more comments. It can also be thrown into a helper function

Reply:
or several helper functions, whatever you feel is right